<a href="https://colab.research.google.com/github/Ali-Shahrez/flyrank-ml-internship/blob/main/work/notebooks/w01_research_question.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# ML-02 — Research Question and Provisional Lane

This skeleton is yours to fill. Work the sections **in order** — each one has a one-line hint. Simple words, honest numbers.

> Working with an AI assistant? Tell it to read `skills/README.md` first and load the one skill this assignment names on its card.

## 1. My lane (or freestyle) and why

*Name your lane — or say 'freestyle' and describe your own question. One short paragraph: why this one?*

I'm choosing Lane 2: Refresh / Content Opportunity Scoring. I already spent
time in the Week 2 starter notebook building a decision tree to flag which
pages should be reviewed first, comparing it against a hand-written rule, and
testing both with a client-holdout split. That hands-on experience showed me
something I wouldn't have expected — the hand-written rule looked great
in-sample but collapsed on new clients, while the tree held up. I want to dig
deeper into why that happened and build a stronger version of this pipeline
for my capstone, rather than starting over in a lane I have no intuition for yet.

## 2. The question: decision, action, cost of a wrong call

*What decision does your work improve? Who acts on it? What does a wrong recommendation cost?*

The question: which pages should a content team prioritize refreshing, expanding,
or leaving alone? The decision-maker is an SEO/content strategist deciding where
to spend limited editing time each week. If the model's recommendation is wrong
in one direction (flagging a healthy page as needing work), the cost is wasted
editing effort on a page that didn't need it. If wrong in the other direction
(missing a genuinely declining page), the cost is a real client losing organic
traffic and rankings that could have been caught early. Both failure directions
have a real cost, which is why precision at the top of the ranked list (Precision@20,
Precision@50) matters more here than overall accuracy.

## 3. Quick look at the data (2-3 real numbers)

*Load the starter CSV below and show 2-3 real numbers that make your lane look worth the next 7 weeks.*

In [2]:
!git clone -q https://github.com/Ali-Shahrez/flyrank-ml-internship.git
%cd flyrank-ml-internship

import pandas as pd
df = pd.read_csv("data/raw/content_refresh_anonymized.csv")
print(f"Loaded {len(df):,} rows")

/content/flyrank-ml-internship
Loaded 30,000 rows


In [3]:
print(f"Total pages: {len(df):,}")
print(f"Unique clients: {df['client_id'].nunique()}")

declining_share = (df['trend_direction'] == 'down').mean()
print(f"Share of pages trending down: {declining_share:.1%}")

print(f"Median avg_position: {df['avg_position'].median():.1f}")
print(f"Median content age (days): {df['content_age_days'].median():.0f}")

Total pages: 30,000
Unique clients: 32
Share of pages trending down: 54.2%
Median avg_position: 10.8
Median content age (days): 236


In [5]:
(df['avg_position'] == 0).sum()

np.int64(1205)

The starter dataset has 30,000 pages across 32 clients. 54.2% of pages are
currently trending down in performance — over half the dataset, which means
decline isn't a rare edge case here; it's close to the norm, giving plenty of
real examples to learn from. The median page sits at position 10.8 (bottom of
page 1 / top of page 2) and is about 236 days old (roughly 8 months). That
combination is telling: most pages aren't ranking poorly from a cold start —
they're mid-pack and aging, which is exactly the profile a refresh/decay model
should be built around, rather than a "how do we break into rankings at all"
problem.

## 4. Careful words: what I can and can't claim

*Write what your work will be able to say (observed, directional, decision-support) — and what it never will (causal proof, 'predicting Google').*

What I can claim: this work is observational and decision-support only. I can
say that certain measurable signals — position, content age, impressions,
engagement — are associated with whether a page's impressions are trending
up or down, and that a model trained on historical client data can rank
pages by estimated risk better than a hand-written rule, based on holdout
precision. Results are directional and useful for prioritizing which pages a
content team reviews first, not a guarantee about any individual page's future.

What I can't claim: I'm not predicting Google's ranking algorithm, and I'm
not proving causation. The starter label, `trend_direction == "down"`, is a
proxy label, not a future outcome — it's a bucket calculated from the current
window (>20% drop in impressions over the most recent 30 days vs. the prior
30), not something measured after a decision point. That means it can't
distinguish real decline from look-alikes: consolidation (a sibling page
absorbing the traffic), seasonality, SERP/AI-driven click loss, or plain
noise from a single volatile month. 54.2% of pages carry this label — close
to a coin flip — so I should describe pages as "flagged by this proxy rule,"
not "confirmed declining." A stronger version of this work would use a
future-window label instead (prior 90 days of features predicting the next
30 days), which is the direction I'd take this in beyond Week 1. Finally, I
can't generalize past the 32 clients in this sample — my own client-holdout
test showed a hand rule that looked accurate on training clients failed on
one it hadn't seen, a caution against overclaiming generalization even when
a model performs well here.

## Self-check

Before you submit, confirm each line honestly:

- [ ] Every section above is filled — markdown thinking AND the code that backs it
- [ ] The notebook runs top to bottom with no errors (Runtime → Run all)
- [ ] No client names, URLs, or private queries anywhere
- [ ] My claims use careful words: observed, measured, directional, decision-support
- [ ] Committed to my repo under `work/notebooks/` — then submit your repo URL on the card. Done.